tools/run_tests/run_xds_tests.py

```diff
@@ -2866,6 +3217,26 @@ def __init__(self, compute, alpha_compute, project, project_num):
                     test_timeout(gcp, backend_service, instance_group)
                 elif test_case == 'fault_injection':
                     test_fault_injection(gcp, backend_service, instance_group)
+                elif test_case == 'api_listener':
+                    if CLIENT_HOSTS:
+                        logger.info('skipping api_listener test case because '
+                                    'client processes on existing client hosts '
+                                    'are out of test scope.')
+                        continue
+                    client_process = test_api_listener(gcp, backend_service, instance_group,
```

Comment:
What happens to the previously started `client_process`? 

Oh, I see it is terminated inside this method, but that's fairly hard to track. Is there any way to restructure things to start the client process with whatever params these tests need in the beginning, so we avoid starting up a (useless?) client_process here just to then terminate it within these methods?

Reply:
> I see it is terminated inside this method, but that's fairly hard to track.

I actually have similar concern.

In the `test_api_listener` test case, it was not because i need a different parameter than the current setting, but to avoid failures in `finally` stage to put back the configurations for the next test.  During test I had to use a new `urlmap+tp+fr` (non `0.0.0.0` and proxyValidation=false) and delete the old `urpmap+tp+fr`. At the end of the test, I have to recover to the a `0.0.0.0` and proxylessValidation=true. I was worried that during that time window, the `0.0.0.0` and original port combination has been taken by someone else to cause the tests to be flaky, which is hard to detect. That is why I decided to repalce with a new client_process that will definitely be successful selecting a new port.
But I think in `test_api_listener` this is fairly less worrisome than in the `forwarding_rule_port_match` test case.

In the `forwarding_rule_port_match` test, we actually make use the original params to test the port not matching case (in [1.a](https://docs.google.com/document/d/1CM3QdAnV0cCd1NiKRj9gSz8Pt38JHPMYme8SJB1fEEw/edit?resourcekey=0-s8rSM-pJenGSrSvSC8EH-A#)), then we terminate the `client_process` and create another one with params we need, then we recover to the original one.
Alternatively for `forwarding_rule_port_match` test case, we can split  ([1.a](https://docs.google.com/document/d/1CM3QdAnV0cCd1NiKRj9gSz8Pt38JHPMYme8SJB1fEEw/edit?resourcekey=0-s8rSM-pJenGSrSvSC8EH-A#) and  [1.bcd](https://docs.google.com/document/d/1CM3QdAnV0cCd1NiKRj9gSz8Pt38JHPMYme8SJB1fEEw/edit?resourcekey=0-s8rSM-pJenGSrSvSC8EH-A#)) and for the latter we can do as you suggested.

Reply:
As I understand it, the client process is only used by the single test case method (e.g., `test_api_listener`). Each test case gets its own client process. So I don't think there is a need to restore or return the "original" client_process when these test methods finish. I think this means you can simplify `test_api_listener` to not accept (or return) a `client_process` parameter, and `forwarding_rule_port_match` doesn't need to return a new `client_process`.

Reply:
It is true that the client process is re-started and then terminated for each test case. Hm I think the problem is in FR port instead of `client_process`, and the `client_process` depends on FR port (to construct `server_uri`). We only have to restore the FR, which may fail for that the specific port that was initialized only once before all the test cases. So we only need to re-pick a FR port among potential port ranges and guarantees to be successful and let the existing code spin up the new client_process for the next test case. Same for the `forwarding_rule_port_match` case that changed the FR port as well.

Reply:
Just replying in case this is waiting on me (if you have already have an update in progress that's fine, I just don't want to be blocking you here) - it sounds like this can be simplified then to stop returning a `client_process` from either of the new test methods. It seems fine to pass `client_process` to the test methods, where needed, but since any returned `client_process` will not be used after the test case returns, it seems simpler to remove the code and logic around reconstructing the `client_process` in the new test cases' `finally` blocks.

Reply:
Yes it looks I can avoid touching `client_process` completely in `api_listener_test`.
But still have to deal with `client_process` in `forwarding_rule_port_match` because within `forwarding_rule_port_match` I tested against original `client_process` and then terminate original `client_process` and spin up a new one for test default port purposes. 
So either we split the test case into multiple, which is the latest change and it seems cleaner to me.
Alternatively we can combine them and I still have return a client_process as I do not want to bypass the existing exception handling between each test cases. And it would be error prone and hard to debug. 
There are tradeoffs. I might be biased so let me know what you think.

Reply:
> it seems simpler to remove the code and logic around reconstructing the client_process in the new test cases' finally blocks.

oh I missed the previous comment.  Passing in `client_process` still seems asking for trouble, but may not be a huge problem. It also needs to bypass the error handling between each test cases. (killing it leaves none-zero return code).

Reply:
Sorry, I missed this last comment. I think just setting the `client_process` in the main loop to None after invoking `forwarding_rule_port_match` would be sufficient to work around the error handling between test. Or create a list of test methods (containing `[forwarding_rule_port_match]`) where `client_process` is not created at all in the main loop. I agree either gets a bit messy...but any approach that avoids creating an entirely new client process that only exists to avoid a non-zero error code seems good enough to me.